### PR TITLE
Move catalog search indexer outside stores loop

### LINKED ISSUE #33984: [Issue] Move catalog search indexer outside stores loop
This issue is automatically created based on existing pull request: magento/magento2#33227: Move catalog search indexer outside stores loop

---------
### Description (*)
There's a plugin (catalogsearchFulltextProductAssignment / Magento\Elasticsearch\Model\Indexer\Fulltext\Plugin\Category\Product\Action\Rows) in the Magento Elastisearch module that hooks in on the Category Products indexer (Magento\Catalog\Model\Indexer\Category\Product\Action\Rows)

This plugin receives the category id's that are reindexed by the Category Products indexer. This plugin collects the assigned products from these categorie id's per store. And inside this loop it triggers a Catalog Search Fulltext reindex for these product id's.

The problem is that the indexer is executed within the stores loop. But the indexer itself already does a reindex for all stores.

We ran in to this issue with our setup that contains 12 stores. 1 category was passed in to the Category Products indexer. The plugin was executed after fetches all connected products for this category (1000+). These 1000 were reindexed in the 12 stores loop and indexed 12 times in each loop.

### Manual testing scenarios (*)
1. Have a Magento setup with a lot of stores
2. Have a category with a lot of products
3. Disable the category, save it, enable the category, save it.
4. Check how long the indexer Category Products takes

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by jasperzeinstra**
### Description (*)
There's a plugin (catalogsearchFulltextProductAssignment / Magento\Elasticsearch\Model\Indexer\Fulltext\Plugin\Category\Product\Action\Rows) in the Magento Elastisearch module that hooks in on the Category Products indexer (Magento\Catalog\Model\Indexer\Category\Product\Action\Rows)

This plugin receives the category id's that are reindexed by the Category Products indexer. This plugin collects the assigned products from these categorie id's per store. And inside this loop it triggers a Catalog Search Fulltext reindex for these product id's.

The problem is that the indexer is executed within the stores loop. But the indexer itself already does a reindex for all stores.

We ran in to this issue with our setup that contains 12 stores. 1 category was passed in to the Category Products indexer. The plugin was executed after fetches all connected products for this category (1000+). These 1000 were reindexed in the 12 stores loop and indexed 12 times in each loop.

### Manual testing scenarios (*)
1. Have a Magento setup with a lot of stores
2. Have a category with a lot of products
3. Disable the category, save it, enable the category, save it.
4. Check how long the indexer Category Products takes

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#33984: Move catalog search indexer outside stores loop

### CODE REVIEW

**Comment by mrtuvn** [comment]
Thanks you for contribution!
Can you cover changes with automation tests ?
Just curiously how this impact with other indexer ? Previously i saw some people claims when they reindexer `catalog_category_products` magento also trigger `catalog_search_index` too. Is this different use case or may related with this one

**Comment by mrtuvn** [comment]
cc: @engcom-Alfa @engcom-Bravo

**Comment by mrtuvn** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php [hunk 1]
```diff
@@ -77,12 +77,15 @@ public function afterExecute(
...
         if (!empty($entityIds) && $indexer->isScheduled()) {
+            $productIds = [];
+
             foreach ($this->storeManager->getStores() as $store) {
                 $indexTable = $this->getIndexTable((int) $store->getId(), $useTempTable);
-                $productIds = $this->getProductIdsFromIndex($indexTable, $entityIds);
-                if (!empty($productIds)) {
-                    $indexer->reindexList($productIds);
-                }
+                $productIds = array_merge($productIds, $this->getProductIdsFromIndex($indexTable, $entityIds));
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php]:
```suggestion
                $productIds[] = $this->getProductIdsFromIndex($indexTable, $entityIds);
```

File: app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php [hunk 2]
```diff
@@ -77,12 +77,15 @@ public function afterExecute(
...
+            }
+
+            if (!empty($productIds)) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php:88]:
```suggestion
            $productIds = array_merge([], ...$productIds);
            if (!empty($productIds)) {
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please don't do array_merge in a loop, usually it's causing performance issues.

File: app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php [hunk 3]
```diff
@@ -77,12 +77,15 @@ public function afterExecute(
...
+                $indexer->reindexList(array_unique($productIds));
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php:89]: Potentially, we might get here few thousand product IDs here. 

Let's split all product ids into some small enough pieces using `array_chunk` (up to 50 products) and reindex this small piece.

**Comment by ihor-sviziev** [comment]
@

**Comment by ihor-sviziev** [comment]
@jasperzeinstra, will you be able to update this PR with suggested changes?

File: app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php (hunk 3 above)
Comment (hostep) [reply app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php:89]: Out of curiosity: what is the benefit of chunking the amount of productId's up? The original code also didn't do this.

Comment (ihor-sviziev) [reply app/code/Magento/Elasticsearch/Model/Indexer/Fulltext/Plugin/Category/Product/Action/Rows.php:89]: I just looked at this index - it seems like it already has a batching mechanism inside, so you're right. We don't need to use array_chunk here.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Unfortunately, it will be too hard to cover this change with automated test

**Comment by kandy** [comment]
@jasperzeinstra @ihor-sviziev  I assume that this change does not change functional behavior, only improves performance. 
so, can we run perf test? Like generate medium profile and run "magento indexer:reindex catalogserach_fulltext" before/after and compare the indexation time

**Comment by ihor-sviziev** [comment]
@kandy, can we generate big amount of stores and products assigned to them?

**Comment by kandy** [comment]
@ihor-sviziev yes, https://devdocs.magento.com/guides/v2.4/config-guide/cli/config-cli-subcommands-perf-data.html

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests EE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9168](https://jira.corp.magento.com/browse/ENGCOM-9168) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @jasperzeinstra , @ihor-sviziev  
Thanks for the collaboration & contribution!

:x: QA not Passed

"magento indexer:reindex catalogsearch_fulltext" delays in its indexation time

**Manual Scenario Steps**

1. Installed a fresh Magento2 instance with sample data.
2. Created 12 different stores with unique store domain and base url.
3. Created categories for each store and added 1000 products to each. (Total 12000+ products got added)
4. Executed the _catalogsearch_fulltext_ indexer command
5. Validated the indexer total execution time.

**:x: Actual result** The indexation time increased by ~30 seconds compared to before fix. The total indexation time taken is 1.43 minutes
>  image

**:heavy_check_mark: Expected result** Before fix, total indexation time was 1.10minutes for above data
> image

**Comment by hostep** [comment]
@engcom-Alfa: I think the comparison should be done between this PR and the parent commit of this branch (being https://github.com/magento/magento2/commit/32fda1cba6ebd1bd7837d01dc34accf67469d560 in this case). Not on the latest 2.4-develop, because there are already more changes done to 2.4-develop and maybe some of them also influence the performance of this indexer.

Not saying that the results will be different, just saying that the comparison is not accurate in my opinion.

**Comment by ihor-sviziev** [comment]
@engcom-Alfa, could you check if you merged the 2.4-develop into the branch with changes from PR before testing?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
As we discussed with @engcom-Alfa in Slack - he tested changes with merged in the recent 2.4-develop branch.

@jasperzeinstra, could you double-check why your changes reducing indexation speed, which is not expected?

**Comment by hostep** [comment]
From what I can tell the testing results still make no sense since the code that was changed is only triggered by the partial reindexing system and not by the full reindexing system. Also running `indexer:reindex` twice on the same database will probably always be faster the second time you execute it, so this is not an accurate way of testing this.

So to properly test this, the performance of the cronjob `indexer_update_all_views` should be tested I think. Easiest way is probably by using [magerun2](https://files.magerun.net/)  by executing `php n98-magerun2.phar sys:cron:run indexer_update_all_views`
But first you'll need to trigger some changes that will make the `catalog_category_product` indexer dirty. Then make a backup of the database, test the performance of the `indexer_update_all_views` cronjob without changes from the PR, restore backup of database, now test with changes from the PR.

Does this make sense?

**Comment by engcom-Alfa** [comment]
> From what I can tell the testing results still make no sense since the code that was changed is only triggered by the partial reindexing system and not by the full reindexing system. Also running `indexer:reindex` twice on the same database will probably always be faster the second time you execute it, so this is not an accurate way of testing this.
> 
> So to properly test this, the performance of the cronjob `indexer_update_all_views` should be tested I think. Easiest way is probably by using [magerun2](https://files.magerun.net/) by executing `php n98-magerun2.phar sys:cron:run indexer_update_all_views`
> So first indexer modes should be set to 'schedule'. Then you'll need to trigger some changes that will make the `catalog_category_product` indexer dirty somehow. Then make a backup of the database, test the performance of the `indexer_update_all_views` cronjob without changes from the PR, restore backup of database, now test with changes from the PR.
> 
> Does this make sense?

Thanks for the inputs @hostep , Yes, I executed full indexing by referring one of the above comment. I will re-do it following your further instructions.
Also I have executed the command twice in same database and the indexing will be faster second time comparing first time. However, I executed the indexing command second time using PR fixed changes. So, second executed result should be very much faster than the first execution time, but it is more!
Anyhow, I will comeback as per your further instructed execution steps soon!
Thanks

**Comment by engcom-Alfa** [comment]
Hi @jasperzeinstra , @ihor-sviziev, @hostep 

:x: QA not Passed Again as per below execution steps as well. 

**Manual Scenario Steps**

1. Created 2 separate Magento instances fresh with sample data (namely- **_prbefore_** & **_prafter_**).
2. Created a new category in each (used exactly same name and configuration in both instances)
3. **_prbefore_** instance is a 2.4-develop instance with latest update; and **_prafter_** instance is having this PR fixes code remotely switched and upgraded.
4. Set both instances indexer to _index by schedule_
5. By importing 10k products and by updating existing data, disturbed the indexing.
6. Executed the command _php n98-magerun2.phar sys:cron:run indexer_update_all_views_


**:heavy_check_mark: Expected result After Fix** 
-  _php n98-magerun2.phar sys:cron:run indexer_update_all_views_ command execution time should be lesser compared to prior fix; Whatever the execution time currently being taken by the cron job, has to be reduced after this PR fix


**:x: Actual result After Fix** 
- _php n98-magerun2.phar sys:cron:run indexer_update_all_views_ command execution time is hardly resulting lesser compared to prior fix. The command execution time taken before the fix was- 19mins57secs; Execution time taken after the fix is- 19mins47secs
- So, reduced time was lesser than 0.01% and it is very much negligible
- Below are the screenshots of the commands executed in both instances.
- image
- image

**Comment by ihor-sviziev** [comment]
@engcom-Alfa, when you should see the bigger difference only in cases when you will edit the **same product(s)** multiple times.

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by engcom-Alfa** [comment]
**Manual testing scenario:**

1. Create a new category in each instances (used exactly same name and configuration in both instances)

2. Set both instances indexer to _index by schedule_
 
3. Import ~1000 products as such. Edit the many products multiple times by their price/quantity/weight as such. (I edited totally 200 products; each product 20 times). Do exactly same in both instances so that data indexer disturbs.

4. Also create a _Marketing>cart price rule_ and _Content>Block_ in both the instances exactly same

5. Execute the command _php n98-magerun2.phar sys:cron:run indexer_update_all_views_ in the _prbefore_ instance and note down the total time taken

6. Execute the command _php n98-magerun2.phar sys:cron:run indexer_update_all_views_ in the _prafter_ instance and note down the total time taken


**Before:** :heavy_multiplication_x: Command execution time taken in the _prbefore_ instance is 6 seconds
image


**After:** :heavy_check_mark:  Command execution time taken in the _prafter_ instance is nearly 4 seconds (performance has increased by 30-35%)
image


This has a specific fix belong to cron job performance issue. There is no impact on any other modules,  also indexing has no any other issue as I tested as per previous comments. Hence there is no regression required on this!

**Pull request merged**